anisotropy versus radial coordinate

stggh opened the pull request:
From issue #176. This is really @DanHickstein's baby, that I somehow hijacked.  It simply allows the determination of the anisotropy parameter for a radial grid, determined from @DanHickstein clever code [example](https://github.com/PyAbel/PyAbel/issues/176#issue-153615099):
```python
r = np.arange(0,300,2)
r_range = list(zip(r[:-1], r[1:]))
```
In this PR I have:
 1. Created new function `abel.tools.vmi.anisotropy()` to provide a single call to obtain the anisotropy parameter. Previously, 2 steps were required, one to evaluate intensity vs angle (for a given radial range) `abel.tools.vmi.radial_integration()`, the second to evaluate the anisotropy parameter from a fit to the intensity variation `abel.tools.vmi.anisotropy_parameter()`. See `examples/example_O2_PES_PAD.py`

 2. Modified `abel.tools.vmi.anisotropy_parameter()` to accept an array of intensity vs thetas
    previously this required a for loop to process the return arrays of `abel.tools.vmi.angular_integration()`. Again, see `examples/example_O2_PES_PAD.py`

 3. (Yet) another example  'example_PAD.py` demonstrating the new function calls
  Individual peaks (red points, with error bars in plot below):
     ```python
     r_range = [(93, 111), (145, 162), (255, 280), (330, 350), (350, 370),
           (370, 390), (390, 410), (410, 430)]

     # anisotropy parameter for each tuple r_range
      Beta, Amp, R = abel.tools.vmi.anisotropy(AIM, r_range)
    ```
    and the whole radial grid (to emulate `basex`/`linbasex`), (green line in plot below):
    ```python
    # OR  anisotropy parameter for ranges (0, 20), (20, 40) ...
    Br, Ar, Rr = abel.tools.vmi.anisotropy(AIM, 20)
    ```

![example_pad](https://cloud.githubusercontent.com/assets/10932229/16575486/ec3b1734-42cb-11e6-8ae4-7cc291de442e.png)

I am not 100% happy with the whole radial-grid output (green line in plot), I would prefer it to look like the output from `basex`/`linbasex`.  I guess  I should plot the anisotropy parameter outputs from `basex`/`linbasex` for comparison. 

Open for comment ...

stggh commented:
Hmm, I am not sure why Travis-CI for Python3.3 fails, all others (including the AppVeyor build) pass.

DanHickstein commented on abel/tools/vmi.py:
```diff
@@ -13,13 +13,13 @@
-    
-    Note: the use of Jacobian=True applies the correct Jacobian for the integration of a 3D object in spherical coordinates.
+
+    Note: the use of Jacobian=True applies the correct Jacobian for the
```
maybe change "correct Jacobian" to "correct Jacobian (r*sin(theta))"

DanHickstein replied:
Oh, this is mentioned in the Parameters section. Disregard.

DanHickstein commented:
Thanks for adding this @stggh! I didn't look to carefully at the code yet, but the general idea seems great.  

The disagreement between the red dots and green curve is just a result of different bin boundaries being used, correct? And we would expect them to be on top of each other given the same bins?

What are you unhappy about regarding the output? BASEX.exe provided one anisotropy point per output pixel, so perhaps you can emulate the old (crappy) behavior by just setting your bin step to 1. In this case, the output is going to be really noisy, so when you plot it, you'll likely need to set some intensity threshold and only plot the anisotropy where you have sufficient signal (set all of the below threshold anisotropy points to np.nan).

stggh commented:
Hmm, a small uniform grid fails dismally, due to the lack of intensity in most regions of the fit:
![anisotropy](https://cloud.githubusercontent.com/assets/10932229/16602405/3509d9e2-4357-11e6-9753-f6f4d8964b7a.png)

In reality, the fixed spaced grid determination of the anisotropy parameter, from a least-squares fit to the intensity variation, has little utility, due to the large error bar associated with fits for low intensity spectral radial regions. It lacks the advantages inherent in the `basex`/`linbasex` approaches. However, this PR does provide a nice interface to extract anisotropy parameters from any image:

```python
Beta, Amp, Rmid = abel.tools.vmi.anisotropy(AIM, radial_ranges=[(rmin0, rmax0), (rmin1, rmax1), ...])
```
which I claim should provide the most accurate anisotropy parameter determination.

I will tidy up `example_PAD.py` which might better be adapted from `example_linbasex_hansenlaw.py` as in the figure above.

DanHickstein commented:
well, the small uniform grid is almost doing the right thing in the regions where there is signal. But linbasex (and the blue dots) really does a nice job.

stggh commented:
I updated `examples/example_PAD.py`

![example_pad](https://cloud.githubusercontent.com/assets/10932229/16645204/aa335a5e-4465-11e6-97da-898ae7a75b05.png)

`linbasex` does a pretty good job determining the anisotropy parameter, and it is nice that it is a by-product of the method.

Q: @DanHickstein, @DhrubajyotiDas  is it possible for `basex` to return the anisotropy parameter variation in the same way as `linbasex`?

I will (hope to!) add something to the docs to discuss the anisotropy parameter evaluation.

For this PR, specifically the code in `abel/tools/vmi.py` it may be better to simply have `radial_integration()` return the anisotropy parameter fits. ie. remove the pseudo `anisotropy()` function and add the line `Beta, Amp = anisotropy_parameter(theta, IvsTheta)` to `radial_integration()`.  I will do this later tonight or early tomorrow.

DanHickstein commented:
This looks really nice @stggh!

Using basex, it should be possible to analytically calculate the anisotropy parameter. However, I'm not sure if the anisotropy parameter would actually be much nicer than what you get from fitting functions to the image as your function can do for any transform method. The reason is that Basex is using a basis set that corresponds to a pixel-by-pixel transform. In contrast, the linbasex method is fitting to functions that are quite smooth in the angular dimension. It's the ideal case for extracting the anisotropy parameters, and I doubt that any method that doesn't fit to polar basis functions like these could really compete for producing nice anisotropy distributions. It's probably worth noting this observation in the PyAbel docs somewhere.

stggh commented:
The image anisotropy parameter determination is now rationalized to the single function `abel.tools.vmi.radial_integration()`. Updated `examples/example_PAD.py` to match. 

```python
Beta, Amp, Rmid, Ivstheta, theta = abel.tools.vmi.radial_integration(HIM.transform, r_range)
```

Added `doc/anisotropy_parameter.rst`.  Note:

 1.  this is photoelectron specific.
 2.  my description of `linbasex` method is vague. 


For consideration:
    
  There is a namespace collision between `linbasex` use of `Beta` and the anisotropy parameter. Currently, `linbasex` returns `Beta[0]` = speed, and `Beta[1]` = anisotropy parameter. I think, strictly speaking, this data "should" be in the array as `Beta[2]` ?  @gerber-t  sorry for the lack of rigour. I will fix this issue later.

stggh commented:
This PR extends the evaluation of the anisotropy parameter &beta; for specific radial regions (usually those that cover strong spectral features) to emulate the full radial determination provided by `linbasex` and `basex`. It achieves this by providing for an integer `step` parameter, which to define a radial grid `(0, step), (step, 2*step), ... , (rmax-step*2, rmax-step)`, using the [clever](https://github.com/PyAbel/PyAbel/issues/176#issue-153615099)  python code of @DanHickstein.

The anisotropy parameter for each radial range is determined by first extracting the (integrated) intensity vs angle, and then fitting  `amp*(1 + beta * P2(cos(theta))` to the intensity, with `beta` and
`amp` as parameters, returned together with their parameter fit errors.

As for all the anisotropy parameter determinations, the most accurate values are determined where the image intensity is strong. Consequently, it turns out, that a uniform `step` grid is of little utility for this method. A more accurate determination of the anisotropy parameter is obtained by limiting the radial (integration) range to spectral features of good intensity.

The `linbasex` method fits Legendre polynomials to each radial (Newton sphere) and consequently directly provides the radial distribution of the anisotropy parameter directly.

As illustrated in the figure [above](https://github.com/PyAbel/PyAbel/pull/178#issuecomment-231002656) the two methods provide complementary information.

Main changes: 

 (1) `radial_ranges` now accepts an integer `step` parameter to define a uniform radial grid over which to extract `I  vs  theta` and determine the anisotropy parameter from a least-squares fit.

 (2) `abel.tools.vmi.radial_integration()` has been rationalized to now return the anisotropy parameter (and parameter fit error), instead of only intensity vs angle for each radial range (which then required a separate call to `abel.tools.vmi.anisotropy_parameter()`).

 (3) Updated examples that involve the anisotropy parameter determination: `example_GUI.py, example_linbasex_hansenlaw.py, example_O2_PES_PAD.py, example_simple_GUI.py`, with a new example, `example_PAD.py`, illustrating the two `PyAbel` methods to determine anisotropy parameters (image above).
   Made improvements to `example_GUI.py` to include `linbasex` transform, and to plot the radial variation of the anisotropy parameter, available for this case.

 (4) New documentation `anisotropy.rst`.

 (5) Added unit test for anisotropy parameter, in `abel/testing/test_tools.py`,

stggh commented:
@PyAbel/developers, I think that this PR is ready for merging. Overall, it is only a small change to PyAbel.

DanHickstein commented:
Thanks for doing this @stggh! Sorry to be absent from the conversation - I've been quite busy for the past few days putting together a conference talk. I'll try to take a look at this in the next few days. Otherwise we can merge as is.

stggh commented:
@DanHickstein no worries.  Take your time. My main constraint is that I will be travelling in 3 weeks time, for a period of 3 weeks.

stggh commented:
@DanHickstein I would like to merge this PR to allow a little time to work more on aspects of the `linebasex` method processing, before I travel.

 I have been playing with the anisotropy determination and it provides more of what a user might expect when `Beta[2]` (or the attribute `.Beta[1]`) is plotted not normalized. e.g.
![beta1](https://cloud.githubusercontent.com/assets/10932229/17157924/24bd6542-53d6-11e6-9e58-d1f47e921e1e.png)

Note, this is only a stylistic change (does not affect the anisotropy parameter), but the anisotropy output may be more familiar to uses of other inverse Abel programs - I think?

DanHickstein commented:
Yup, go ahead and merge! Thanks @stggh!
